src/legacy/core_plugins/region_map/public/components/region_map_options.tsx

```diff
@@ -0,0 +1,263 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useEffect, useState, useCallback, useMemo } from 'react';
+import { EuiIcon, EuiLink, EuiPanel, EuiSpacer, EuiText, EuiTitle } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+
+import { toastNotifications } from 'ui/notify';
+import { FileLayerField, VectorLayer, ServiceSettings } from 'ui/vis/map/service_settings';
+import { VisOptionsProps } from 'ui/vis/editors/default';
+import { SelectOption } from '../../../kbn_vislib_vis_types/public/controls/select';
+import { SwitchOption } from '../../../kbn_vislib_vis_types/public/controls/switch';
+import { NumberInputOption } from '../../../kbn_vislib_vis_types/public/controls/number_input';
+import { ORIGIN } from '../../../tile_map/common/origin';
+import { WmsOptions } from '../../../tile_map/public/components/wms_options';
+import { mapToLayerWithId } from '../util';
+import { RegionMapVisParams } from '../types';
+import { RegionMapsConfig } from '../plugin';
+
+const mapLayerForOption = ({ layerId, name }: VectorLayer) => ({
+  text: name,
+  value: layerId,
+});
+
+const mapFieldForOption = ({ description, name }: FileLayerField) => ({
+  text: description,
+  value: name,
+});
+
+export type RegionMapOptionsProps = {
+  serviceSettings: ServiceSettings;
+  includeElasticMapsService: RegionMapsConfig['includeElasticMapsService'];
+} & VisOptionsProps<RegionMapVisParams>;
+
+function RegionMapOptions(props: RegionMapOptionsProps) {
+  const { includeElasticMapsService, serviceSettings, stateParams, vis, setValue } = props;
+  const [vectorLayers, setVectorLayers] = useState<VectorLayer[]>(
+    vis.type.editorConfig.collections.vectorLayers
+  );
+  const [vectorLayerOptions, setVectorLayerOptions] = useState(vectorLayers.map(mapLayerForOption));
+  const currentLayerId = stateParams.selectedLayer && stateParams.selectedLayer.layerId;
+  const fieldOptions = useMemo(
+    () =>
+      ((stateParams.selectedLayer && stateParams.selectedLayer.fields) || []).map(
+        mapFieldForOption
+      ),
+    [currentLayerId]
+  );
+
+  const setEmsHotLink = useCallback(async (layer: VectorLayer) => {
+    const emsHotLink = await serviceSettings.getEMSHotLink(layer);
+    setValue('emsHotLink', emsHotLink);
+  }, []);
+
+  const setLayer = useCallback(
+    async (paramName: 'selectedLayer', value: VectorLayer['layerId']) => {
+      const newLayer = vectorLayers.find(({ layerId }: VectorLayer) => layerId === value);
+
+      if (newLayer) {
+        setValue(paramName, newLayer);
+        setValue('selectedJoinField', newLayer.fields[0]);
+        setEmsHotLink(newLayer);
+      }
+    },
+    [vectorLayers]
```

Comment:
nit: should also include `setValue`

Reply:
Currently we build the logic with static `setValue` prop, but it could be changed in the future.. so it would make sense to watch it also. Thnx!